File: module-infra/src/main/java/com/hanghae/cache/RedisCacheManager.java

package com.hanghae.cache;

import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.RedisTemplate;

import java.time.Duration;
import java.util.Optional;

@RequiredArgsConstructor
public class RedisCacheManager<K, V> implements CacheManager<K, V> {

    private static final Duration DEFAULT_EXPIRE_AFTER_MINUTES = Duration.ofMinutes(10);
    private final RedisTemplate<K, V> cache;

    @Override
    public void put(K key, V value) {
        cache.opsForValue()
                .set(key, value, DEFAULT_EXPIRE_AFTER_MINUTES);
    }

    @Override
    public void put(K key, V value, Duration expireAfter) {
        cache.opsForValue()
                .set(key, value, expireAfter);
    }

    @Override
    public Optional<V> get(K key) {
        return Optional.ofNullable(cache.opsForValue().get(key));
    }

    @Override
    public void remove(K key) {
        cache.delete(key);
    }
}

File: infrastructure/src/main/kotlin/com/hanghe/redis/cache/RedisCacheManager.kt
package com.hanghe.redis.cache

import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.StringRedisTemplate
import java.time.Duration
import kotlin.reflect.KClass

class RedisCacheManager(
    private val redisTemplate: StringRedisTemplate,
    private val objectMapper: ObjectMapper
): CacheManager {

    override fun <T : Any> getOrNull(key: String, clazz: KClass<T>): T? {
        return redisTemplate.opsForValue().get(key)?.let {
            objectMapper.readValue(it, clazz.java)
        }
    }

    override fun <T : Any> getOrPut(key: String, ttl: Duration, clazz: KClass<T>, cacheable: () -> T): T {
        return getOrNull(key, clazz) ?: run {
            put(key, ttl, clazz, cacheable)
        }
    }

    override fun <T : Any> put(key: String, ttl: Duration, clazz: KClass<T>, cacheable: () -> T): T {
        val value = cacheable()

        runCatching {
            val serializedValue = objectMapper.writeValueAsString(value)
            redisTemplate.opsForValue().set(key, serializedValue, ttl)
        }.onFailure { e ->
            logger.error("Failed to cache data put in Redis, key: `$key`, Error Message: ${e.message}")
        }

        return value
    }

    companion object {
        private val logger = LoggerFactory.getLogger(RedisCacheManager::class.java)
    }
}
